File: Phase-5/frontend/src/types/task.ts
export type Priority = "high" | "medium" | "low";
export type RecurringInterval = "daily" | "weekly" | "monthly";
export type TaskStatus = "pending" | "completed";
export type SortBy = "date" | "priority" | "name";
export type SortOrder = "asc" | "desc";

export interface Task {
  id: string;
  user_id: string;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: Priority;
  tags: string[];
  due_date: string | null;
  recurring: RecurringInterval | null;
  reminders: string[];
  created_at: string;
  updated_at: string;
}

export interface TaskFilters {
  status?: TaskStatus | "all";
  priority?: Priority;
  tag?: string;
  search?: string;
  sort_by?: SortBy;
  sort_order?: SortOrder;
}

File: Phase-5/backend/app/models/task.py
from datetime import datetime
from typing import Optional, List
from sqlmodel import Field, SQLModel, Column
from sqlalchemy import String, DateTime, Index
from sqlalchemy.dialects.postgresql import ARRAY


class Task(SQLModel, table=True):
    __tablename__ = "tasks"
    __table_args__ = (
        Index("ix_tasks_tags", "tags", postgresql_using="gin"),
    )

    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: str = Field(index=True)
    title: str = Field(max_length=200)
    description: Optional[str] = Field(default=None)
    status: str = Field(default="pending", index=True)
    priority: str = Field(default="medium", index=True)
    tags: List[str] = Field(
        default=[],
        sa_column=Column(ARRAY(String), nullable=False, server_default="{}"),
    )
    due_date: Optional[datetime] = Field(
        default=None,
        sa_column=Column(DateTime(timezone=True), nullable=True, index=True),
    )
    recurring: Optional[str] = Field(default=None)
    reminders: List[datetime] = Field(
        default=[],
        sa_column=Column(
            ARRAY(DateTime(timezone=True)), nullable=False, server_default="{}"
        ),
    )
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)
